drivers/serial/uart_stm32.c

```diff
@@ -679,45 +726,124 @@ static int uart_stm32_fifo_fill(const struct device *dev,
 	/* Lock interrupts to prevent nested interrupts or thread switch */
 	key = irq_lock();
 
-	while ((size - num_tx > 0) &&
-	       LL_USART_IsActiveFlag_TXE(config->usart)) {
+	while ((size - num_tx > 0) && LL_USART_IsActiveFlag_TXE(config->usart)) {
 		/* TXE flag will be cleared with byte write to DR|RDR register */
 
-		/* Send a character (8bit , parity none) */
-		LL_USART_TransmitData8(config->usart, tx_data[num_tx++]);
+		/* Send a character */
+		fill_fn(config, tx_data, num_tx);
+		num_tx++;
 	}
 
 	irq_unlock(key);
 
 	return num_tx;
 }
 
-static int uart_stm32_fifo_read(const struct device *dev, uint8_t *rx_data,
-				  const int size)
+static void uart_stm32_fifo_fill_with_u8(const struct uart_stm32_config *config,
+					 const void *tx_data, const uint8_t offset)
+{
+	const uint8_t *data = (const uint8_t *)tx_data;
+	/* Send a character (8bit) */
+	LL_USART_TransmitData8(config->usart, data[offset]);
+}
+
+static int uart_stm32_fifo_fill(const struct device *dev, const uint8_t *tx_data, int size)
+{
+	if (uart_stm32_ll2cfg_databits(uart_stm32_get_databits(dev), uart_stm32_get_parity(dev)) ==
+	    UART_CFG_DATA_BITS_9) {
+		return -ENOTSUP;
+	}
+	return uart_stm32_fifo_fill_visitor(dev, (const void *)tx_data, size,
+					    uart_stm32_fifo_fill_with_u8);
+}
+
+typedef void (*fifo_read_fn)(const struct uart_stm32_config *config, void *rx_data,
+			     const uint8_t offset);
+
+static int uart_stm32_fifo_read_visitor(const struct device *dev, void *rx_data, const int size,
+					fifo_read_fn read_fn)
 {
 	const struct uart_stm32_config *config = dev->config;
 	uint8_t num_rx = 0U;
 
-	while ((size - num_rx > 0) &&
-	       LL_USART_IsActiveFlag_RXNE(config->usart)) {
+	while ((size - num_rx > 0) && LL_USART_IsActiveFlag_RXNE(config->usart)) {
 		/* RXNE flag will be cleared upon read from DR|RDR register */
 
-		/* Receive a character (8bit , parity none) */
-		rx_data[num_rx++] = LL_USART_ReceiveData8(config->usart);
+		read_fn(config, rx_data, num_rx);
+		num_rx++;
 
 		/* Clear overrun error flag */
 		if (LL_USART_IsActiveFlag_ORE(config->usart)) {
 			LL_USART_ClearFlag_ORE(config->usart);
-		/*
-		 * On stm32 F4X, F1X, and F2X, the RXNE flag is affected (cleared) by
-		 * the uart_err_check function call (on errors flags clearing)
-		 */
+			/*
+			 * On stm32 F4X, F1X, and F2X, the RXNE flag is affected (cleared) by
+			 * the uart_err_check function call (on errors flags clearing)
+			 */
 		}
 	}
 
 	return num_rx;
 }
 
+static void uart_stm32_fifo_read_with_u8(const struct uart_stm32_config *config, void *rx_data,
```

Comment:
Please apply the same naming convention than for poll functions

Reply:
Fixed